keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,445 @@
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### PodsReady workload condition
+
+First, we introduce a new workload condition, called `PodsReady`, to indicate
+that the first batch of a workload's pods is scheduled, more precisely:
+`job.status.ready + job.status.succeeded >= job.spec.parallelism`.
+
+The `PodsReady` condition is addded to the workload by the Kueue's Job
+reconciler in reaction to a status update of the corresponding Job. Note that,
+verifying if the condition should be added does not require an extra API call as
+the Kueue's Job reconciller already fetches the latest Job object at the
+beginning of the `Reconcile` function. Also note, there is no new condition
+introduced to the Job API.
+
+### Workload API
+
+We extend the Workload API with the new `startupRequirements` field to
+enable the new behavior.
+
+```golang
+// StartupRequirements describes workload startup requirements
+type StartupRequirements string
+
+const (
+  // QuotaAndPodsReady means that workload requires all pods to be running at
+  // the same time
+  StartupRequirementsQuotaAndPodsReady StartupRequirements = "QuotaAndPodsReady"
+  // QuotaOnly means that the only requirement for a workload to start is to
+  // get resources from the cluster quota
+  StartupRequirementsQuotaOnly StartupRequirements = "QuotaOnly"
+)
+
+// WorkloadSpec defines the desired state of Workload
+type WorkloadSpec struct {
+  ...
+  // StartupRequirements describes the workload requirement on the resource allocation.
+  // Possible values are:
+  // - QuotaAndPodsReady: means that workload requires allocation of resources
+  //   to support running of all pods at the same time,
+  // - QuotaOnly: means that the workload does not require all pods to be running
+  //   at the same time.
+  // Defaults to QuotaOnly.
+  StartupRequirements *string `json:"startupRequirements,omitempty"`
+}
+```
+
+### Job annotation
+
+We introduce a new Job annotation, called `kueue.x-k8s.io/startup-requirements`,
+as a JobAPI counterpart of the `startupRequirements` Workload API field. When
+the annotation is present, its value is used as the field value during workload
+creation. Unknown value of the annotation results in leaving the field
+as null.
+
+### The mechanism
+
+When a workload with `startupRequirements=QuotaAndPodsReady` is nominated for admission,
+kueue scheduler blocks its admission (and admission of other workloads nominated
```

Comment:
Workloads nominated in this cycle may belong to different clusterQueues, then no need to block each other since they're isolated, no?

Reply:
Workloads in different cluster queues still use the same cluster, and so may use the same cluster resources underneath, unless using flavors corresponding to disjoint resources, but there is no guarantee on this.

Reply:
I understand that under these scenarios 
- workloads in the same clusterQueue
- clusterQueues under the the same cohort

we may have resource deadlock for they're sharing with each other, but for clusterQueues under different cohorts, the assumption is that clusterQueues are fine divided by administrator(maybe we should better document that one node should only belong to one resource flavor and one clusterQueue), so I think we should take them as resource isolated, then there's no need to avoid coscheduling between them. cc @alculquicondor @ahg-g for thoughts.

But I agree with startup with a strict and simple implementation anyway.